### PR TITLE
Add an option to deny packages or groups of packages

### PR DESCRIPTION
**Pull request by adrienpessu**
# TL;DR;

The goal is to check if someone is adding specific packages or group of packages in the codebase that isn't supposed to be added.

# Context

GitHub customers want to deny access to certain dependencies base on a wide range of criteria other than vulnerabilities and licences. 
Some examples of criteria : 
- The project isn't maintained any more
- We don't want to depend on our competitor's open source libraries
- We don't want any dependencies of a particular ecosystem because it has a bad reputation (security wise), it depends on an entity based in a country we need to avoid for legal reasons, ...

Since those criteria are wide, the simplest way is to deny based on the name or the group of the packages.

# Actions

- [x] Add option for a list of denied packages
- [x] Add option for a list of denied groups of packages
- [x] Add a specific message into the summary
- [x] check the denied lists and failed if found
- [x] Add tests
- [x] Edit Readme

### CODE REVIEW

**Comment by febuiles** [comment]
@adrienpessu thank you for the additional information! From a code POV this looks fine, some things that would be appreciated before merging and releasing a new version:

- [ ] Add an example configuration to: https://github.com/actions/dependency-review-action/blob/main/docs/examples.md and then link to it from the `README` changes. This helps us educate users and create release notes.
- [ ] Create a public repository that uses this action and link to a public run of the Action pointing to your branch (i.e. link to a run of your changes). Example: https://github.com/future-funk/expert-octo-garbanzo/actions/runs/5397412934/job/14612005965?pr=2. This helps us make sure that the code has been tested end to end.

Once these two are done I'll cut the new release.

**Comment by adrienpessu** [comment]
@febuiles I have added the example and I prepared the testing repo : https://github.com/adrienpessu/sample-dependency-review-deny-packages/actions/runs/5794591201?pr=1, I can link it as is, but it uses my forked action. Once merge, I will be able to change it to the real action.
Second question, where do I put the link to the run so people can find it, I didn't see other links in the current repo

**Comment by febuiles** [comment]
@adrienpessu Sorry for the confusion, the test run is for us devs to leave a record here of a successful run of the changes since we don't have proper E2E tests in Actions. We don't need to share it as part of the documentation.

Thanks for the new examples, they point out a change we need to make: We should use PURLs instead of package names. It avoids conflicts for multiple dependencies with the same name but different ecosystems (think of a library called `json` that exists in both your `Gemfile` and your `package.json`). Apologies for not calling out this earlier, I thought we were already using PURLs.

**Comment by adrienpessu** [comment]
@febuiles while I like the flexibility of not having to specify the build system, I understand the need to avoid confusion. 
I made the change and here is an example : https://github.com/adrienpessu/sample-dependency-review-deny-packages/actions/runs/5798862598?pr=1

File: README.md [hunk 1]
```diff
@@ -66,18 +66,20 @@ jobs:
...
+| `comment-summary-in-pr`         | Enable or disable reporting the review summary as a comment in the pull request. If enabled, you must give the workflow or job permission `pull-requests: write`.                 | `true`, `false`                                                                                | `false`       |
+| `deny-packages`                 | Contains a list of denied package's URL.                                                                                                                                          | Any packages [purl](https://github.com/package-url/purl-spec) without the version              | empty         |
+| `deny-groups`                   | Contains a list of denied groups package's name.                                                                                                                                  | Any packages [purl](https://github.com/package-url/purl-spec) without the version and the name | empty         |
```
Comment (febuiles) [inline README.md:82]:
```suggestion
| `deny-packages`                 | Any number of packages to block in a PR.                                                                                                                                           | Package(s) in [purl](https://github.com/package-url/purl-spec) format            | empty         |
| `deny-groups`                   | Any number of groups (namespaces) to block in a PR.                                                                                                                                   | Namespace(s) in [purl](https://github.com/package-url/purl-spec) format (no package name, no version number) | empty         |
```

File: action.yml [hunk 2]
```diff
@@ -47,6 +47,12 @@ inputs:
   comment-summary-in-pr:
     description: A boolean to determine if the report should be posted as a comment in the PR itself. Setting this to true requires you to give the workflow the write permissions for pull-requests
     required: false
+  deny-packages:
+    description: A comma-separated list of packages to deny (e.g. "pkg:npm/express, pkg:pip/pycrypto")
```
Comment (febuiles) [inline action.yml:51]:
```suggestion
    description: A comma-separated list of package URLs to deny (e.g. "pkg:npm/express, pkg:pip/pycrypto")
```

File: action.yml [hunk 3]
```diff
@@ -47,6 +47,12 @@ inputs:
...
+    required: false
+  deny-groups:
+    description: A comma-separated list of groups of packages to deny (e.g. "pkg:npm/express, pkg:pip/pycrypto")
```
Comment (febuiles) [inline action.yml:54]:
```suggestion
    description: A comma-separated list of package URLs for group(s)/namespace(s) to deny (e.g. "pkg:npm/express, pkg:pip/pycrypto")
```

File: docs/examples.md [hunk 4]
```diff
@@ -230,3 +230,31 @@ jobs:
           comment-summary-in-pr: true
           license-check: false
 ```
+
+## Exclude dependencies from their name or groups
+
+Using the `deny-packages` you can exclude dependencies by their full package name. You can add multiple values separated by a comma.
```
Comment (febuiles) [inline docs/examples.md:236]:
```suggestion
Using the `deny-packages` option you can exclude dependencies by their PURL. You can add multiple values separated by a commas.
```

File: docs/examples.md [hunk 5]
```diff
@@ -230,3 +230,31 @@ jobs:
...
+Using the `deny-groups` you can exclude dependencies by their package group name. You can add multiple values separated by a comma.
```
Comment (febuiles) [inline docs/examples.md:237]:
```suggestion
Using the `deny-groups` option you can exclude dependencies by their group name/namespace. You can add multiple values separated by a comma.
```

**Review by febuiles — APPROVED** [review]

**Comment by febuiles** [comment]
@adrienpessu I've merged the changes with some tweaks to the docs. I expect to do a release later today/tomorrow. Thank you very much!

**Comment by adrienpessu** [comment]
Great thank you

**Comment by febuiles** [comment]
By merging I forgot that the `README` changes would be public instantly, so I decided to cut the new release now. Changes are available starting with 3.0.7. Thanks again!

**Pull request merged**